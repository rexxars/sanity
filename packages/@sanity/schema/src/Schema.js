// @flow
import type {SchemaDef, Type} from './flowtypes'

import * as types from './types'
import {createHelpfulError} from './helpfulError'
import inspect from 'object-inspect'
import {compileRegistry} from './compileRegistry'
import {validateSchemaDef} from './validation/validateSchema'

export default {
  compile: function (schemaDef: SchemaDef) {
    if (!(schemaDef && typeof schemaDef === 'object' && Array.isArray(schemaDef.types) && schemaDef.name && typeof schemaDef.name === 'string')) {
      throw createHelpfulError(
        `Expected schema to be an object with a name and an array of types, instead got ${inspect(schemaDef)}`,
        'schema-compile-invalid-arguments'
      )
    }

    const validation = validateSchemaDef(schemaDef, types)

    const registry = compileRegistry(schemaDef, types)

    return {
      name: schemaDef.name,
      get(name: string): Type {
        return registry[name] && registry[name].get()
      },

      has(name: string): boolean {
        return name in registry
      },
      getTypeNames(): string[] {
        return Object.keys(registry)
      }
    }
  }
}
