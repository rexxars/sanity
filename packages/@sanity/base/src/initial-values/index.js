export {resolveInitialValue} from './resolve'
export {getTemplates} from './templates'
