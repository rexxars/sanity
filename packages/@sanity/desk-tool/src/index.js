import React from 'react'
import Icon from 'part:@sanity/base/view-column-icon'
import {route} from 'part:@sanity/base/router'
import DeskTool from './DeskTool'
import {parsePanesSegment} from './utils/parsePanesSegment'
import UUID from '@sanity/uuid'

function maybeRemapStringSegment(segment) {
  return typeof segment === 'string' ? {id: segment} : segment
}

function encodeSegment({id, params}) {
  const parts = params ? [id, JSON.stringify(params)] : [id]
  return parts.join(',')
}

function toState(pathSegment) {
  return parsePanesSegment(decodeURIComponent(pathSegment))
}

function toPath(panes) {
  return panes
    .map(maybeRemapStringSegment)
    .map(encodeSegment)
    .map(encodeURIComponent)
    .join(';')
}

const state = {activePanes: []}

function setActivePanes(panes) {
  state.activePanes = panes
}

function DeskToolPaneStateSyncer(props) {
  return <DeskTool {...props} onPaneChange={setActivePanes} />
}

function getIntentState(intentName, params, currentState) {
  const paneSegments = (currentState && currentState.panes) || []
  const activePanes = state.activePanes || []
  const editDocumentId = params.id || UUID()
  const isTemplate = intentName === 'create' && params.template

  // Loop through open panes and see if any of them can handle the intent
  for (let i = activePanes.length - 1; i >= 0; i--) {
    const pane = activePanes[i]
    if (pane.canHandleIntent && pane.canHandleIntent(intentName, params)) {
      const paneParams = isTemplate ? {template: params.template} : undefined
      return {panes: paneSegments.slice(0, i).concat({id: editDocumentId, params: paneParams})}
    }
  }

  return getFallbackIntentState({documentId: editDocumentId, intentName, params})
}

function getFallbackIntentState({documentId, intentName, params}) {
  const editDocumentId = documentId
  const isTemplateCreate = intentName === 'create' && params.template

  return isTemplateCreate
    ? {editDocumentId, template: params.template}
    : {editDocumentId, type: params.type || '*'}
}

export default {
  router: route('/', [
    route('/edit/:type/:editDocumentId'),
    route({
      path: '/:panes',
      // Legacy URLs, used to handle redirects
      children: [route('/:action', route('/:legacyEditDocumentId'))],
      transform: {
        panes: {toState, toPath}
      }
    })
  ]),
  canHandleIntent(intentName, params) {
    return (
      (intentName === 'edit' && params.id) ||
      (intentName === 'create' && params.type) ||
      (intentName === 'create' && params.template)
    )
  },
  getIntentState,
  title: 'Desk',
  name: 'desk',
  icon: Icon,
  component: DeskToolPaneStateSyncer
}
