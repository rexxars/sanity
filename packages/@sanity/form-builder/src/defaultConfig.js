import {
  ArrayInput,
  BooleanInput,
  EmailInput,
  NumberInput,
  ObjectInput,
  StringInput,
  Reference,
  Image,
  File,
  Slug,
  Any,
  TextInput,
  UrlInput
} from './defaultInputComponents'

import ArrayOfStrings from './inputs/Array/ArrayOfStrings'
import SearchableStringSelect from './inputs/SearchableStringSelect'
import StringSelect from './inputs/StringSelect'
import ArrayOfStringsSelect from './inputs/Array/ArrayOfStringsSelect'
import BlockEditor from './inputs/BlockEditor-slate'

const typeNameToInputMap = {
  object: ObjectInput,
  array: ArrayInput,
  boolean: BooleanInput,
  number: NumberInput,
  text: TextInput,
  email: EmailInput,
  url: UrlInput,
  image: Image,
  any: Any,
  file: File,
  reference: Reference,
  string: StringInput,
  slug: Slug
}

function isArrayOfStrings(type) {
  return type.name === 'array' && type.of.every(ofType => ofType.name === 'string')
}

function hasBlockMember(type) {
  return type.name === 'array'
    && type.of.find(memberType => memberType.name === 'block')
}

function hasListInOptions(type) {
  return type.options && type.options.list
}

function isList(type) {
  return type.options && type.options.list
}

function isSearchable(type) {
  return type.options.searchable
}

function resolveInputComponent(type) {

  // Schema provides predefines list
  if (hasListInOptions(type) && isArrayOfStrings(type)) {
    return ArrayOfStringsSelect
  }

  // Special component for array of strings
  if (isArrayOfStrings(type)) {
    return ArrayOfStrings
  }

  // Special component for array of strings
  if (hasBlockMember(type)) {
    return BlockEditor
  }

  // String input with a select
  if (type.name === 'string' && isList(type)) {
    return isSearchable(type) ? SearchableStringSelect : StringSelect
  }

  return typeNameToInputMap[type.name] || typeNameToInputMap[type.jsonType]
}

function resolvePreviewComponent(type) {
  // leave empty for now
}

export default {
  resolveInputComponent: resolveInputComponent,
  resolvePreviewComponent: resolvePreviewComponent
}
