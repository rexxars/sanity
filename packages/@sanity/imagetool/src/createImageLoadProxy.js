import PropTypes from 'prop-types'
import React from 'react'

export default function createImageLoadProxy(
  Component,
  {error: ErrorComponent, loader: LoaderComponent} = {}
) {
  return class ImageLoadProxy extends React.Component {
    static displayName = `${Component.displayName ||
      Component.name ||
      '<Anonymous>'}$ImageLoadProxy`
    static propTypes = {
      src: PropTypes.string.isRequired
    }

    state = {
      image: null,
      error: null
    }

    componentWillMount() {
      this.loadImage(this.props.src)
    }

    loadImage(src) {
      const image = new Image()
      this.setState({
        image: null,
        error: null
      })

      image.onload = () => {
        this.setState({
          image: image,
          error: null
        })
      }

      image.onerror = () => {
        this.setState({
          error: new Error(`Could not load image from ${JSON.stringify(this.props.src)}`)
        })
      }

      image.src = src
    }

    componentWillReceiveProps(nextProps) {
      if (nextProps.src !== this.props.src) {
        this.loadImage(nextProps.src)
      }
    }

    render() {
      const {error, image} = this.state

      if (!error && !image) {
        // loading
        return LoaderComponent ? <LoaderComponent {...this.props} /> : null
      }
      if (error) {
        return ErrorComponent ? <ErrorComponent {...this.props} error={error} /> : null
      }
      return <Component {...this.props} image={image} />
    }
  }
}
