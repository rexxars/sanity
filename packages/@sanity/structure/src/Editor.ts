import {camelCase} from 'lodash'
import {EditorNode, SerializeOptions, Serializable, StructureNode} from './StructureNodes'
import {getTemplateById} from '@sanity/initial-value-templates'
import {SerializeError, HELP_URL} from './SerializeError'
import {SchemaType} from './parts/Schema'
import {InitialValueTemplateItem, InitialValueTemplateItemBuilder} from './InitialValueTemplateItem'

export type PartialEditorNode = {
  id?: string
  title?: string
  options?: {
    id?: string
    type?: string
    template?: string
  }
  parameters?: {
    [key: string]: any
  }
  initialValueTemplates?: (InitialValueTemplateItem | InitialValueTemplateItemBuilder)[]
}

export class EditorBuilder implements Serializable {
  protected spec: PartialEditorNode

  constructor(spec?: EditorNode) {
    this.spec = spec ? spec : {}
  }

  id(id: string): EditorBuilder {
    return this.clone({id})
  }

  getId() {
    return this.spec.id
  }

  title(title: string) {
    return this.clone({title, id: this.spec.id || camelCase(title)})
  }

  getTitle() {
    return this.spec.title
  }

  documentId(documentId: string): EditorBuilder {
    // Let's try to be a bit helpful and assign an ID from document ID if none is specified
    const paneId = this.spec.id || documentId
    return this.clone({
      id: paneId,
      options: {
        ...(this.spec.options || {}),
        id: documentId
      }
    })
  }

  getDocumentId() {
    return this.spec.options && this.spec.options.id
  }

  schemaType(documentType: SchemaType | string): EditorBuilder {
    return this.clone({
      options: {
        ...(this.spec.options || {}),
        type: typeof documentType === 'string' ? documentType : documentType.name
      }
    })
  }

  getSchemaType() {
    return this.spec.options && this.spec.options.type
  }

  initialValueTemplates(templates: InitialValueTemplateItem[]) {
    return this.clone({
      options: {
        ...(this.spec.options || {})
      },
      initialValueTemplates: templates
    })
  }

  getInitialValueTemplates() {
    return this.spec.initialValueTemplates
  }

  serialize({path = [], index, hint}: SerializeOptions = {path: []}): EditorNode {
    const urlId = path[index || path.length - 1]
    const {initialValueTemplates} = this.spec

    // Try to grab document ID / editor ID from URL if not defined
    const id = this.spec.id || (urlId && `${urlId}`)
    const options = {id, type: undefined, template: undefined, ...this.spec.options}

    if (typeof id !== 'string' || !id) {
      throw new SerializeError('`id` is required for editor nodes', path, index, hint).withHelpUrl(
        HELP_URL.ID_REQUIRED
      )
    }

    if (!options || !options.id) {
      throw new SerializeError(
        'document id (`id`) is required for editor nodes',
        path,
        id,
        hint
      ).withHelpUrl(HELP_URL.DOCUMENT_ID_REQUIRED)
    }

    return {
      ...this.spec,
      id,
      type: 'document',
      options: {id: options.id, type: options.type, template: options.template},
      initialValueTemplates:
        initialValueTemplates &&
        initialValueTemplates.map(item => (isBuilder(item) ? item.serialize() : item))
    }
  }

  clone(withSpec: PartialEditorNode = {}) {
    const builder = new EditorBuilder()
    const options = {...(this.spec.options || {}), ...(withSpec.options || {})}
    builder.spec = {...this.spec, ...withSpec, options}
    return builder
  }
}

function isBuilder(item: Serializable | StructureNode): item is Serializable {
  return typeof (item as Serializable).serialize === 'function'
}

export function editorWithInitialValueTemplate(
  templateId: string,
  parameters?: {[key: string]: any}
) {
  const template = getTemplateById(templateId)
  if (!template) {
    throw new Error(`Template with ID "${templateId}" not defined`)
  }

  return new EditorBuilder()
    .schemaType(template.schemaType)
    .initialValueTemplates([
      {id: templateId, templateId, parameters, type: 'initialValueTemplateItem'}
    ])
}
