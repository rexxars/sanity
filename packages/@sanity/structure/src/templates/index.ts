import {Template, TemplateBuilder} from './Template'
import {Schema, SchemaType, getDefaultSchema} from '../parts/Schema'

function defaultTemplateForType(
  schemaType: string | SchemaType,
  sanitySchema?: Schema
): TemplateBuilder {
  let type: SchemaType
  if (typeof schemaType === 'string') {
    const schema = sanitySchema || getDefaultSchema()
    type = schema.get(schemaType)
  } else {
    type = schemaType
  }

  return new TemplateBuilder({
    id: type.name,
    schemaType: type.name,
    title: type.title || type.name,
    icon: type.icon,
    value: type.initialValue || {_type: type.name}
  })
}

function defaults(sanitySchema?: Schema) {
  const schema = sanitySchema || getDefaultSchema()
  return schema
    .getTypeNames()
    .map(typeName => defaultTemplateForType(schema.get(typeName), sanitySchema))
}

export default {
  template: (spec?: Template) => new TemplateBuilder(spec),
  defaults,
  defaultTemplateForType
}
