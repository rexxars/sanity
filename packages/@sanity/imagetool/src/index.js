import React from 'react'
import ImageTool from './ImageTool'
import createImageLoadProxy from './createImageLoadProxy'

/* eslint-disable react/no-multi-comp, react/prop-types */
export default createImageLoadProxy(ImageTool, {
  loading(props) {
    return <div>Loading...</div>
  },
  error(props) {
    return (
      <div>
        Could not load image "{props.src}" {props.error.message}
      </div>
    )
  }
})
