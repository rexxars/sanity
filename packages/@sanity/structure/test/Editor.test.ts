import {StructureBuilder as S} from '../src'
import {getDefaultSchema} from '../src/parts/Schema'

test('builds editor node through constructor', () => {
  expect(
    S.editor({
      id: 'foo',
      title: 'some title',
      options: {
        id: 'docId',
        type: 'book'
      }
    }).serialize()
  ).toMatchSnapshot()
})

test('throws on missing id', () => {
  expect(() =>
    S.editor()
      .documentId('docId')
      .schemaType('book')
      .serialize()
  ).toThrowError(/`id` is required/)
})

test('infers ID from title if not specified', () => {
  expect(
    S.editor()
      .title('Hei der')
      .getId()
  ).toEqual('heiDer')
  expect(
    S.editor()
      .id('zing')
      .title('Hei der')
      .getId()
  ).toEqual('zing')
  expect(
    S.editor()
      .title('Hei der')
      .id('blah')
      .getId()
  ).toEqual('blah')
})

test('throws on missing document ID', () => {
  expect(() =>
    S.editor()
      .id('id')
      .title('title')
      .schemaType('book')
      .serialize()
  ).toThrowError(/document id/)
})

test('can construct with schema type instead of schema type name', () => {
  expect(
    S.editor()
      .schemaType(getDefaultSchema().get('post'))
      .id('yeah')
      .title('Yeah')
      .documentId('wow')
      .serialize()
  ).toMatchSnapshot()
})

test('can construct using builder', () => {
  expect(
    S.editor()
      .id('yeah')
      .title('Yeah')
      .documentId('wow')
      .schemaType('book')
      .serialize()
  ).toMatchSnapshot()
})

test('can construct using builder (alt)', () => {
  expect(
    S.editor()
      .schemaType('book')
      .id('yeah')
      .title('Yeah')
      .documentId('wow')
      .serialize()
  ).toMatchSnapshot()
})

test('builder is immutable', () => {
  const original = S.editor()
  expect(original.id('foo')).not.toEqual(original)
  expect(original.title('foo')).not.toEqual(original)
  expect(original.documentId('moo')).not.toEqual(original)
  expect(original.schemaType('author')).not.toEqual(original)
})

test('getters work', () => {
  const original = S.editor()
  expect(original.id('foo').getId()).toEqual('foo')
  expect(original.title('bar').getTitle()).toEqual('bar')
  expect(original.documentId('moo').getDocumentId()).toEqual('moo')
  expect(original.schemaType('author').getSchemaType()).toEqual('author')
})
