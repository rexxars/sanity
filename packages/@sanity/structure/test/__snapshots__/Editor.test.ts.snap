// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds editor node through constructor 1`] = `
Object {
  "id": "foo",
  "options": Object {
    "id": "docId",
    "template": undefined,
    "type": "book",
  },
  "parameters": Object {
    "foo": "bar",
  },
  "title": "some title",
  "type": "document",
}
`;

exports[`can construct using builder (alt) 1`] = `
Object {
  "id": "yeah",
  "options": Object {
    "id": "wow",
    "template": "developer",
    "type": "book",
  },
  "parameters": Object {
    "bar": "foo",
  },
  "title": "Yeah",
  "type": "document",
}
`;

exports[`can construct using builder 1`] = `
Object {
  "id": "yeah",
  "options": Object {
    "id": "wow",
    "template": undefined,
    "type": "book",
  },
  "parameters": Object {
    "foo": "bar",
  },
  "title": "Yeah",
  "type": "document",
}
`;

exports[`can construct with schema type instead of schema type name 1`] = `
Object {
  "id": "yeah",
  "options": Object {
    "id": "wow",
    "template": undefined,
    "type": "post",
  },
  "title": "Yeah",
  "type": "document",
}
`;
